src/fs.cpp

```diff
@@ -90,6 +90,122 @@ bool FileLock::TryLock()
+*  |             +     +                  w                  |
+*  |         +                            r                  |
+*  |         +   +                        r+                 |
+*  |         +   +     +                  w+                 |
+*  |         +   +           +            a+                 |
+*  |         +               +            a+                 |
+*  +---------------------------------------------------------+
+*  |   +         +                        wb                 |
+*  |   +         +           +            ab                 |
+*  |   +                     +            ab                 |
+*  |   +         +     +                  wb                 |
+*  |   +     +                            rb                 |
+*  |   +     +   +                        r+b                |
+*  |   +     +   +     +                  w+b                |
+*  |   +     +   +           +            a+b                |
+*  |   +     +               +            a+b                |
+*  +---------------------------------------------------------+
+*/
+static std::string openmodeToStr(std::ios_base::openmode mode)
+{
+    switch (mode & ~std::ios_base::ate) {
+    case std::ios_base::out:
+    case std::ios_base::out | std::ios_base::trunc:
+        return "w";
+    case std::ios_base::out | std::ios_base::app:
+    case std::ios_base::app:
+        return "a";
+    case std::ios_base::in:
+        return "r";
+    case std::ios_base::in | std::ios_base::out:
+        return "r+";
+    case std::ios_base::in | std::ios_base::out | std::ios_base::trunc:
+        return "w+";
+    case std::ios_base::in | std::ios_base::out | std::ios_base::app:
+    case std::ios_base::in | std::ios_base::app:
+        return "a+";
+    case std::ios_base::out | std::ios_base::binary:
+    case std::ios_base::out | std::ios_base::trunc | std::ios_base::binary:
+        return "wb";
+    case std::ios_base::out | std::ios_base::app | std::ios_base::binary:
+    case std::ios_base::app | std::ios_base::binary:
+        return "ab";
+    case std::ios_base::in | std::ios_base::binary:
+        return "rb";
+    case std::ios_base::in | std::ios_base::out | std::ios_base::binary:
+        return "r+b";
+    case std::ios_base::in | std::ios_base::out | std::ios_base::trunc | std::ios_base::binary:
+        return "w+b";
+    case std::ios_base::in | std::ios_base::out | std::ios_base::app | std::ios_base::binary:
+    case std::ios_base::in | std::ios_base::app | std::ios_base::binary:
+        return "a+b";
+    default:
+        return std::string();
+    }
+}
+
+void ifstream::open(const fs::path& p, std::ios_base::openmode mode)
+{
+    close();
+    file = fopen(p, openmodeToStr(mode).c_str());
+    if (file == nullptr) {
+        return;
+    }
+    filebuf = std::move(__gnu_cxx::stdio_filebuf<char>(file, mode));
+    rdbuf(&filebuf);
+    if (mode & std::ios_base::ate) {
+        seekg(0, std::ios_base::end);
```

Comment:
Would be nice to have a unit test to make sure this logic works. I believe this code will run on travis.